File: face/.agents/skills/twitter/scripts/post_x.py
import undetected_chromedriver as uc
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time
import sys
import json
import os

def main():
    if len(sys.argv) < 2:
        print('Usage: python3 post_x_uc.py "Tweet Text"')
        sys.exit(1)
        
    text = sys.argv[1]

    # Load secrets
    secrets_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../../secrets.json'))
    try:
        with open(secrets_path, 'r') as f:
            secrets = json.load(f)
            username = secrets['twitter']['username']
            password = secrets['twitter']['password']
            email = secrets['twitter']['email']
    except Exception as e:
        print(f"FAILED TO LOAD SECRETS: {e}")
        sys.exit(1)

    print('🦞 [UC-Stealth] Launching undetected Chrome browser...')
    
    # Needs to be headless to run in background.
    # UC handles headless masking, but we use the specific UC `headless_new=True` flag or standard `headless=True`.
    # `headless=True` in UC removes most fingerprinting vulnerabilities.
    options = uc.ChromeOptions()
    options.add_argument('--no-sandbox')
    options.headless = True
    
    driver = uc.Chrome(options=options)
    
    try:
        print('   [UC-Stealth] Navigating to X.com login...')
        driver.get("https://x.com/i/flow/login")
        wait = WebDriverWait(driver, 15)
        
        print('   [UC-Stealth] Waiting for username field...')
        user_input = wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, 'input[autocomplete="username"]')))
        
        print('   [UC-Stealth] Entering username...')
        user_input.send_keys(username)
        
        # Click Next
        next_button = driver.find_element(By.XPATH, '//span[text()="Next"]/ancestor::button | //button[contains(., "Next")]')
        next_button.click()
        
        # Unusual activity check
        try:
            email_wait = WebDriverWait(driver, 5)
            email_input = email_wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, 'input[name="text"]')))
            print('   [UC-Stealth] Unusual activity detected, entering email...')
            email_input.send_keys(email)
            next_button = driver.find_element(By.XPATH, '//span[text()="Next"]/ancestor::button | //button[contains(., "Next")]')
            next_button.click()
        except:
            pass # No unusual activity
            
        print('   [UC-Stealth] Waiting for password field...')
        password_input = wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, 'input[name="password"]')))
        print('   [UC-Stealth] Entering password...')
        password_input.send_keys(password)
        
        login_button = driver.find_element(By.XPATH, '//span[text()="Log in"]/ancestor::button | //button[@data-testid="LoginForm_Login_Button"]')
        login_button.click()
        
        print('   [UC-Stealth] Waiting for timeline to load...')
        wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, '[data-testid="SideNav_NewTweet_Button"], [data-testid="tweetButtonInline"], [data-testid="AppTabBar_Home_Link"]')))
        
        print('   [UC-Stealth] Navigating to compose window...')
        driver.get('https://x.com/compose/post')
        
        wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, '[data-testid="tweetTextarea_0"]')))
        print('   [UC-Stealth] Typing tweet...')
        
        tweet_box = driver.find_element(By.CSS_SELECTOR, '[data-testid="tweetTextarea_0"]')
        # Click first to focus, then send keys
        tweet_box.click()
        time.sleep(1)
        tweet_box.send_keys(text)
        
        time.sleep(1)
        
        print('   [UC-Stealth] Clicking Post...')
        post_btn = driver.find_element(By.CSS_SELECTOR, '[data-testid="tweetButton"]')
        post_btn.click()
        
        time.sleep(4)
        print('   [UC-Stealth] ✅ Successfully posted to X.com using Python Undetected_Chromedriver.')

    except Exception as e:
        print(f"   [UC-Stealth] ❌ Posting failed: {e}")
        driver.save_screenshot('uc_error.png')
        with open('error_uc_dom.html', 'w') as f:
            f.write(driver.page_source)
        sys.exit(1)
    finally:
        driver.quit()

if __name__ == "__main__":
    main()

File: face/.agents/skills/twitter/scripts/post_x.js
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

(async () => {
    const text = process.argv[2];
    if (!text) {
        console.error('Usage: node post_x_cdp.js "Tweet text"');
        process.exit(1);
    }

    const secretsPath = '../../../../secrets.json';
    const secrets = JSON.parse(fs.readFileSync(secretsPath, 'utf8'));
    const username = secrets.twitter.username;
    const password = secrets.twitter.password;
    const email = secrets.twitter.email;

    // Read Pinchtab's active Chrome debugger port
    const devToolsFile = '/Users/zerbytheboss/.pinchtab/chrome-profile/DevToolsActivePort';
    if (!fs.existsSync(devToolsFile)) {
        console.error('Pinchtab Chrome is not running! Please start pinchtab first.');
        process.exit(1);
    }

    const port = fs.readFileSync(devToolsFile, 'utf8').split('\n')[0].trim();
    const cdpUrl = `http://127.0.0.1:${port}`;

    console.log(`🦞 [CDP-Bridge] Hijacking Pinchtab Browser via ${cdpUrl}...`);

    // Connect Playwright to the running Pinchtab browser instance
    const browser = await chromium.connectOverCDP(cdpUrl);

    // Get the default context
    const contexts = browser.contexts();
    const context = contexts.length > 0 ? contexts[0] : await browser.newContext();
    const page = context.pages().length > 0 ? context.pages()[0] : await context.newPage();

    try {
        console.log('   [CDP-Bridge] Navigating to X.com login...');
        await page.goto('https://x.com/i/flow/login', { waitUntil: 'domcontentloaded' });

        // Helper to forcefully mutate React state bindings
        const fillReactInput = async (selector, value) => {
            await page.evaluate(({ sel, val }) => {
                let input = document.querySelector(sel);
                if (!input) return;
                let lastValue = input.value;
                input.value = val;
                let event = new Event('input', { bubbles: true });
                event.simulated = true;
                let tracker = input._valueTracker;
                if (tracker) tracker.setValue(lastValue);
                input.dispatchEvent(event);
            }, { sel: selector, val: value });
            await page.waitForTimeout(500);
        };

        console.log('   [CDP-Bridge] Waiting for username field...');
        await page.waitForSelector('input[autocomplete="username"]', { timeout: 15000 });
        console.log('   [CDP-Bridge] Entering email...');
        await fillReactInput('input[autocomplete="username"]', email);
        await page.locator('button:has-text("Next")').click({ force: true });

        try {
            await page.waitForTimeout(2000);
            const unusual = await page.$('input[name="text"]');
            if (unusual) {
                console.log('   [CDP-Bridge] Unusual activity detected, entering username...');
                await fillReactInput('input[name="text"]', username);
                await page.locator('button:has-text("Next")').first().click({ force: true });
            }
        } catch (e) {
            // Normal flow
        }

        console.log('   [CDP-Bridge] Waiting for password field...');
        try {
            await page.waitForSelector('input[name="password"]', { timeout: 10000 });
        } catch (pwError) {
            console.error('   [CDP-Bridge] Failed to find password field. Dumping page text...');
            const pageText = await page.evaluate(() => document.body.innerText);
            fs.writeFileSync('error_text_dump.txt', pageText);
            throw pwError;
        }
        console.log('   [CDP-Bridge] Entering password...');
        await fillReactInput('input[name="password"]', password);
        await page.locator('[data-testid="LoginForm_Login_Button"]').click({ force: true });

        console.log('   [CDP-Bridge] Waiting for timeline to load...');
        await page.waitForSelector('[data-testid="SideNav_NewTweet_Button"], [data-testid="tweetButtonInline"], [data-testid="AppTabBar_Home_Link"]', { timeout: 20000 });

        console.log('   [CDP-Bridge] Navigating directly to compose link...');
        await page.goto('https://x.com/compose/post', { waitUntil: 'load' });

        await page.waitForSelector('[data-testid="tweetTextarea_0"]', { timeout: 15000 });
        console.log('   [CDP-Bridge] Typing tweet...');
        await page.type('[data-testid="tweetTextarea_0"]', text, { delay: 50 });

        await page.waitForTimeout(1000); // Let React register the value

        console.log('   [CDP-Bridge] Clicking Post...');
        await page.locator('[data-testid="tweetButton"]').click();

        await page.waitForTimeout(5000);
        console.log('   [CDP-Bridge] ✅ Successfully posted to X.com using hijacked Pinchtab session.');

    } catch (error) {
        console.error('   [CDP-Bridge] ❌ Posting failed:', error);
        await page.screenshot({ path: 'cdp_error.png' });
        process.exit(1);
    } finally {
        await browser.close();
    }
})();
